// @paytabs/react-native-paytabs/src/PaymentSDKTheme.js
export default class PaymentSDKTheme {
  constructor(
    logoImage,
    primaryColor,
    primaryFontColor,
    primaryFont,
    secondaryColor,
    secondaryFontColor,
    secondaryFont,
    strokeColor,
    strokeThickness,
    inputsCornerRadius,
    buttonColor,
    buttonFontColor,
    buttonFont,
    titleFontColor,
    titleFont,
    backgroundColor,
    placeholderColor
  ) {
    this.logoImage = logoImage;
    this.primaryColor = primaryColor;
    this.primaryFontColor = primaryFontColor;
    this.primaryFont = primaryFont;
    this.secondaryColor = secondaryColor;
    this.secondaryFontColor = secondaryFontColor;
    this.secondaryFont = secondaryFont;
    this.strokeColor = strokeColor;
    this.strokeThinckness = strokeThickness;
    this.inputsCornerRadius = inputsCornerRadius;
    this.buttonColor = buttonColor;
    this.buttonFontColor = buttonFontColor;
    this.buttonFont = buttonFont;
    this.titleFontColor = titleFontColor;
    this.titleFont = titleFont;
    this.backgroundColor = backgroundColor;
    this.placeholderColor = placeholderColor;
  }
}
