// react-native-sticky-parallax-header/lib/module/predefinedComponents/DetailsHeader/components/HeaderBar.js
import * as React from 'react';
import { Pressable, StyleSheet } from 'react-native';
import Animated, { useAnimatedStyle } from 'react-native-reanimated';
import { SafeAreaView } from 'react-native-safe-area-context';
import { colors, commonStyles } from '../../../constants';
import IconRenderer from '../../common/components/IconRenderer';
import { parseAnimatedColorProp } from '../../common/utils/parseAnimatedColorProp';
const HIT_SLOP = {
  top: 15,
  left: 15,
  bottom: 15,
  right: 15
};
export const HeaderBar = ({
  backgroundColor,
  enableSafeAreaTopInset,
  headerTitleContainerAnimatedStyle,
  leftTopIcon,
  leftTopIconAccessibilityLabel,
  leftTopIconOnPress,
  leftTopIconTestID,
  rightTopIcon,
  rightTopIconAccessibilityLabel,
  rightTopIconOnPress,
  rightTopIconTestID,
  title,
  titleStyle,
  titleTestID = 'DetailsHeaderBarTitleTestID'
}) => {
  const wrapperAnimatedStyle = useAnimatedStyle(() => {
    return {
      // TypeScript complains about AnimatedNode<StyleProp<ViewStyle>> from reanimated v1
      backgroundColor: parseAnimatedColorProp(backgroundColor)
    };
  }, [backgroundColor]);
  const safeAreaEdges = ['left', 'right'];

  if (enableSafeAreaTopInset) {
    safeAreaEdges.push('top');
  }

  return /*#__PURE__*/React.createElement(SafeAreaView, {
    edges: safeAreaEdges,
    style: commonStyles.container
  }, /*#__PURE__*/React.createElement(Animated.View, {
    style: [commonStyles.headerWrapper, wrapperAnimatedStyle]
  }, leftTopIcon ? /*#__PURE__*/React.createElement(Pressable, {
    accessibilityLabel: leftTopIconAccessibilityLabel,
    accessibilityRole: "button",
    hitSlop: HIT_SLOP,
    onPress: leftTopIconOnPress,
    style: styles.leftHeaderButton,
    testID: leftTopIconTestID
  }, /*#__PURE__*/React.createElement(IconRenderer, {
    icon: leftTopIcon
  })) : null, /*#__PURE__*/React.createElement(Animated.View, {
    style: [styles.headerTitleContainer, headerTitleContainerAnimatedStyle]
  }, /*#__PURE__*/React.createElement(Animated.Text, {
    style: [styles.headerTitle, titleStyle],
    testID: titleTestID
  }, title)), rightTopIcon ? /*#__PURE__*/React.createElement(Pressable, {
    accessibilityLabel: rightTopIconAccessibilityLabel,
    accessibilityRole: "button",
    hitSlop: HIT_SLOP,
    onPress: rightTopIconOnPress,
    style: styles.rightHeaderButton,
    testID: rightTopIconTestID
  }, /*#__PURE__*/React.createElement(IconRenderer, {
    icon: rightTopIcon
  })) : null));
};
const styles = StyleSheet.create({
  headerTitle: {
    color: colors.white,
    fontSize: 16,
    lineHeight: 20,
    paddingHorizontal: 12,
    textAlign: 'left'
  },
  headerTitleContainer: {
    alignItems: 'flex-start',
    flex: 8
  },
  leftHeaderButton: {
    alignItems: 'flex-start',
    flex: 1
  },
  rightHeaderButton: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
