// rn-zendesk/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_native_1 = require("react-native");
const { RNZendesk } = react_native_1.NativeModules;
var ZDKHelpCenterOverviewGroupType;
(function (ZDKHelpCenterOverviewGroupType) {
    ZDKHelpCenterOverviewGroupType[ZDKHelpCenterOverviewGroupType["DEFAULT"] = 0] = "DEFAULT";
    ZDKHelpCenterOverviewGroupType[ZDKHelpCenterOverviewGroupType["SECTION"] = 1] = "SECTION";
    ZDKHelpCenterOverviewGroupType[ZDKHelpCenterOverviewGroupType["CATEGORY"] = 2] = "CATEGORY";
})(ZDKHelpCenterOverviewGroupType = exports.ZDKHelpCenterOverviewGroupType || (exports.ZDKHelpCenterOverviewGroupType = {}));
/**
 * initialize zendesk with provided configuration
 *
 * @see https://developer.zendesk.com/embeddables/docs/ios_support_sdk/sdk_initialize#initializing-the-support-sdk-required
 * @param config appId, cliendId, zendeskUrl
 */
function initialize(config) {
    RNZendesk.initialize(config);
}
exports.initialize = initialize;
/**
 * set user details
 * Note: zendesk should be initialized before accessing `identifyAnonymous`
 *
 * @param name name of the user
 * @param email email of the user
 */
function identifyAnonymous(name, email) {
    RNZendesk.identifyAnonymous(name, email);
}
exports.identifyAnonymous = identifyAnonymous;
/**
 * launch help center
 * Note: zendesk should be initialized before accessing `showHelpCenter`
 *
 * @param options
 */
function showHelpCenter(options) {
    RNZendesk.showHelpCenter(options);
}
exports.showHelpCenter = showHelpCenter;
