// react-native-sticky-parallax-header/lib/module/predefinedComponents/DetailsHeader/DetailsHeaderScrollView.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from 'react';
import { View } from 'react-native';
import Animated, { useAnimatedStyle } from 'react-native-reanimated';
import { commonStyles } from '../../constants';
import { StickyHeaderScrollView } from '../../primitiveComponents/StickyHeaderScrollView';
import { parseAnimatedColorProp } from '../common/utils/parseAnimatedColorProp';
import { HeaderBar } from './components/HeaderBar';
import { useDetailsHeader } from './hooks/useDetailsHeader';
export const DetailsHeaderScrollView = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    backgroundColor,
    children,
    contentContainerStyle,
    decelerationRate = 'fast',
    enableSafeAreaTopInset = true,
    leftTopIcon,
    leftTopIconAccessibilityLabel,
    leftTopIconOnPress,
    leftTopIconTestID,
    nestedScrollEnabled = true,
    overScrollMode = 'never',
    renderHeader,
    renderHeaderBar,
    rightTopIcon,
    rightTopIconAccessibilityLabel,
    rightTopIconOnPress,
    rightTopIconTestID,
    scrollEventThrottle = 16,
    title,
    titleStyle,
    ...rest
  } = props;
  const {
    headerTitleContainerAnimatedStyle,
    onMomentumScrollEnd,
    onScroll,
    onScrollEndDrag,
    renderHeader: defaultRenderHeader,
    scrollViewRef
  } = useDetailsHeader(props);
  React.useImperativeHandle(ref, () => scrollViewRef.current);
  const wrapperAnimatedStyle = useAnimatedStyle(() => {
    return {
      backgroundColor: parseAnimatedColorProp(backgroundColor)
    };
  }, [backgroundColor]);
  return /*#__PURE__*/React.createElement(Animated.View, {
    style: [commonStyles.container, wrapperAnimatedStyle]
  }, renderHeaderBar ? renderHeaderBar() : /*#__PURE__*/React.createElement(HeaderBar, {
    backgroundColor: backgroundColor,
    enableSafeAreaTopInset: enableSafeAreaTopInset,
    headerTitleContainerAnimatedStyle: headerTitleContainerAnimatedStyle,
    leftTopIcon: leftTopIcon,
    leftTopIconAccessibilityLabel: leftTopIconAccessibilityLabel,
    leftTopIconOnPress: leftTopIconOnPress,
    leftTopIconTestID: leftTopIconTestID,
    rightTopIcon: rightTopIcon,
    rightTopIconAccessibilityLabel: rightTopIconAccessibilityLabel,
    rightTopIconOnPress: rightTopIconOnPress,
    rightTopIconTestID: rightTopIconTestID,
    title: title,
    titleStyle: titleStyle
  }), /*#__PURE__*/React.createElement(View, {
    style: commonStyles.container
  }, /*#__PURE__*/React.createElement(StickyHeaderScrollView, _extends({
    ref: scrollViewRef
  }, rest, {
    contentContainerStyle: contentContainerStyle,
    decelerationRate: decelerationRate,
    nestedScrollEnabled: nestedScrollEnabled,
    onMomentumScrollEnd: onMomentumScrollEnd,
    onScroll: onScroll,
    onScrollEndDrag: onScrollEndDrag,
    overScrollMode: overScrollMode,
    renderHeader: renderHeader !== null && renderHeader !== void 0 ? renderHeader : defaultRenderHeader,
    scrollEventThrottle: scrollEventThrottle
  }), children)));
});
