// react-native-sticky-parallax-header/lib/module/predefinedComponents/DetailsHeader/withDetailsHeaderFlashList.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import * as React from 'react';
import { View } from 'react-native';
import Animated, { useAnimatedStyle } from 'react-native-reanimated';
import { commonStyles } from '../../constants';
import { withStickyHeaderFlashList } from '../../primitiveComponents/withStickyHeaderFlashList';
import { parseAnimatedColorProp } from '../common/utils/parseAnimatedColorProp';
import { HeaderBar } from './components/HeaderBar';
import { useDetailsFlashListHeader } from './hooks/useDetailsFlashListHeader';
export function withDetailsHeaderFlashList(flashListComponent) {
  const StickyHeaderFlashList = withStickyHeaderFlashList(flashListComponent);
  return /*#__PURE__*/React.forwardRef((props, ref) => {
    const {
      backgroundColor,
      decelerationRate = 'fast',
      enableSafeAreaTopInset = true,
      leftTopIcon,
      leftTopIconAccessibilityLabel,
      leftTopIconOnPress,
      leftTopIconTestID,
      nestedScrollEnabled = true,
      overScrollMode = 'never',
      renderHeader,
      renderHeaderBar,
      rightTopIcon,
      rightTopIconAccessibilityLabel,
      rightTopIconOnPress,
      rightTopIconTestID,
      scrollEventThrottle = 16,
      title,
      titleStyle,
      ...rest
    } = props;
    const {
      headerTitleContainerAnimatedStyle,
      renderHeader: defaultRenderHeader,
      scrollViewRef,
      onScroll,
      onScrollEndDrag,
      onMomentumScrollEnd
    } = useDetailsFlashListHeader(props);
    React.useImperativeHandle(ref, () => scrollViewRef.current);
    const wrapperAnimatedStyle = useAnimatedStyle(() => {
      return {
        backgroundColor: parseAnimatedColorProp(backgroundColor)
      };
    }, [backgroundColor]);
    return /*#__PURE__*/React.createElement(Animated.View, {
      style: [commonStyles.container, wrapperAnimatedStyle]
    }, renderHeaderBar ? renderHeaderBar() : /*#__PURE__*/React.createElement(HeaderBar, {
      backgroundColor: backgroundColor,
      enableSafeAreaTopInset: enableSafeAreaTopInset,
      headerTitleContainerAnimatedStyle: headerTitleContainerAnimatedStyle,
      leftTopIcon: leftTopIcon,
      leftTopIconAccessibilityLabel: leftTopIconAccessibilityLabel,
      leftTopIconOnPress: leftTopIconOnPress,
      leftTopIconTestID: leftTopIconTestID,
      rightTopIcon: rightTopIcon,
      rightTopIconAccessibilityLabel: rightTopIconAccessibilityLabel,
      rightTopIconOnPress: rightTopIconOnPress,
      rightTopIconTestID: rightTopIconTestID,
      title: title,
      titleStyle: titleStyle
    }), /*#__PURE__*/React.createElement(View, {
      style: commonStyles.container
    }, /*#__PURE__*/React.createElement(StickyHeaderFlashList, _extends({
      ref: scrollViewRef
    }, rest, {
      decelerationRate: decelerationRate,
      nestedScrollEnabled: nestedScrollEnabled,
      onScroll: onScroll,
      onMomentumScrollEnd: onMomentumScrollEnd,
      onScrollEndDrag: onScrollEndDrag,
      overScrollMode: overScrollMode,
      renderHeader: renderHeader !== null && renderHeader !== void 0 ? renderHeader : defaultRenderHeader,
      scrollEventThrottle: scrollEventThrottle
    }))));
  });
}
